### DISCUSSION TITLE
How to copy files onto an hda image from the linux command line?

### DISCUSSION THREAD

**Discussion by TorZidan** [akuker/RASCSI#1059] [Q&A]
I want to be able copy a file (e.g. "DaynaPORT 7.5.3.sea", downloaded from macintoshgarden.org or from the web) and place it into the root folder of an "hda" image file (if there is enough disk space on it).
Would like to do this from the linux command line, e.g.
```
hdacopy DaynaPORT 7.5.3.sea MacHD-50MB.hda
```
It should be able to properly set the Resource Fork (File Type and Creator attributes) on the target file, for popular file extensions.

Is there such utility? Or is there an easier way to bring mac software files from the web onto an hda image?

**Comment by hilburger** [comment]
> I want to be able copy a file (e.g. "DaynaPORT 7.5.3.sea", downloaded from macintoshgarden.org or from the web) and place it into the root folder of an "hda" image file (if there is enough disk space on it). Would like to do this from the linux command line, e.g.
> 
> ```
> hdacopy DaynaPORT 7.5.3.sea MacHD-50MB.hda
> ```
> 
> It should be able to properly set the Resource Fork (File Type and Creator attributes) on the target file, for popular file extensions.
> 
> Is there such utility? Or is there an easier way to bring mac software files from the web onto an hda image?

The way I have found so far is on Windows and not the most elegant but maybe it helps you: 
#1063

**Reply by rdmark** [reply]
This method will only work if Windows is able to make sense of the HFS file system on the image file that you mount. I think there are software packages out there that add HFS support to Windows, but I haven't played around with them myself.

**Comment by rdmark** [comment]
The Linux kernel has support for HFS and HFS+ -- https://www.kernel.org/doc/html/latest/filesystems/hfs.html

I would suggest you play around with mounting the file system and copy files over once it's mounted.

Any tool with the functionality that you propose would probably involve automating the mounting and dismounting the file system in this fashion.

Personally, I would suggest you play around with the PiSCSI web interface's "Download to ISO" feature, or run Netatalk for AppleShare file sharing. The former only requires the Mac to have a CD-ROM driver, which shouldn't be too hard to set up.

**Reply by TorZidan** [reply]
I played with the linux kernel HFS capabilities a while back (on ubuntu ver. 21), and found out that the 1st attempt to write to the mounted image succeeds, but it corrupts the image file: subsequent attempts to unmount + mount that image lets you mount it in read-only mode only, which is a dead end for me.

For people interested trying it:
```
sudo apt-get install --yes hfsprogs
sudo modprobe -a hfs
lsmod | grep hfs
sudo mkdir /mnt/PiSCSI
sudo mount -t hfs -o loop,offset=49152 ~/PiScsi-Basilisk2/PiSCS-1GB.hda /mnt/PiSCSI
(see [https://ubuntuforums.org/archive/index.php/t-1576011.htm](https://ubuntuforums.org/archive/index.php/t-1576011.html)l for info on how to find the correct offset value above)
mount | grep PiSCSI
ls -la /mnt/PiSCSI
If the mount was successful, you will see the mount information, and get a directory listing of the contents from your hard disk image.
To copy files onto the image:  sudo cp <filename> /mnt/PiSCSI
To unmount:    sudo umount /mnt/PiSCSI
```
